opentreemap/works_management/tests.py

```diff
@@ -74,15 +79,82 @@ def test_work_order(self):
-        t = make_task(self.instance, self.user,
-                      make_plot(self.instance, self.user))
+        t = make_task(self.instance, self.user, self.plot)
         with self.assertRaises(UserTrackingException):
             t.save()
+
+    def test_instance_sequence_helpers(self):
+        self.assertEqual(1, self.instance.get_next_task_sequence(9))
+        self.assertEqual(10, self.instance.get_next_task_sequence())
+
+        self.assertEqual(1, self.instance.get_next_work_order_sequence(9))
+        self.assertEqual(10, self.instance.get_next_work_order_sequence())
+
+        # Try to commit bad data.
+        self.instance.task_sequence_number = 0
+        self.instance.work_order_sequence_number = 0
+        self.instance.save()
+        self.instance.refresh_from_db()
+
+        self.assertEqual(11, self.instance.task_sequence_number)
+        self.assertEqual(11, self.instance.work_order_sequence_number)
+
+    def test_work_order_sequence(self):
+        w1 = make_work_order(self.instance, self.user)
+        w2 = make_work_order(self.instance, self.user)
+        self.assertEqual(1, w1.reference_number)
+        self.assertEqual(2, w2.reference_number)
+
+        # Sequence numbers should be unique by instance.
+        other_instance = make_instance()
+        w1 = make_work_order(other_instance, self.user)
+        w2 = make_work_order(other_instance, self.user)
+        self.assertEqual(1, w1.reference_number)
+        self.assertEqual(2, w2.reference_number)
+
+    def test_task_sequence(self):
+        t1 = make_task(self.instance, self.user, self.plot)
+        t2 = make_task(self.instance, self.user, self.plot)
+        self.assertEqual(1, t1.reference_number)
+        self.assertEqual(2, t2.reference_number)
+
+        # Sequence numbers should be unique by instance.
+        other_instance = make_instance()
+        t1 = make_task(other_instance, self.user, self.plot)
+        t2 = make_task(other_instance, self.user, self.plot)
+        self.assertEqual(1, t1.reference_number)
+        self.assertEqual(2, t2.reference_number)
+
+    def test_work_order_sequence_bulk_create(self):
+        orders = [make_work_order(self.instance, self.user, save=False)
+                  for _ in range(10)]
+
+        value = self.instance.get_next_work_order_sequence(len(orders))
+        for i in range(len(orders)):
+            orders[i].reference_number = value + i
+
+        WorkOrder.objects.bulk_create(orders)
+
+        self.instance.refresh_from_db()
+        self.assertEqual(11, self.instance.work_order_sequence_number)
+
+    def test_task_sequence_bulk_create(self):
+        tasks = [make_task(self.instance, self.user, self.plot, save=False)
+                 for _ in range(10)]
+
+        value = self.instance.get_next_task_sequence(len(tasks))
+        for i in range(len(tasks)):
+            tasks[i].reference_number = value + i
+
+        Task.objects.bulk_create(tasks)
```

Comment:
What if `WorkOrder` and `Task` each had a custom `Manager` that overrode `bulk_create` to do the three lines of code above?